Docs: request security (#30937)
grafana/grafana: docs/sources/enterprise/request-security.md

+++
title = "Request security"
description = "Grafana Enterprise request security"
keywords = ["grafana", "security", "enterprise"]
weight = 110
+++

# Request security

> > **Note:** Available in Grafana Enterprise v7.4+.

Reviewer: ```suggestion
> **Note:** Available in Grafana Enterprise v7.4 and later versions.
```

Reviewer 2: @xlson, please apply this format throughout.

> Request security introduces ways to limit requests from the Grafana server and focuses on requests generated by users.

Reviewer 3: ```suggestion
Request security introduces ways to limit requests from the Grafana server and focuses on requests generated by users.
```

What does "focuses" mean here? "Request security focuses on ..."

Author: I'm trying to signal that the focus is requests that are a direct result of user input. Datasource config, alerting. This feature doesn't cover all requests, like the request to the database or the ldap system. Those are done when setting up grafana and are not covered. I'm having a hard time finding the right way to make this distinction.

Reviewer 3: Thank you for the added context. How do you like this alternative?

Reviewer 3: My gut says that a human can focus, or make a machine focus, rather than a machine focusing (read concentrating) on something.

Author: Makes sense to me, thanks!

> Request security introduces ways to limit requests from the Grafana server, and it targets requests that are generated by users.

Reviewer: ```suggestion
Request security introduces new ways to limit requests from the Grafana server, and it targets requests that are generated by users.
```

Reviewer 3: Please avoid "new", which gets old really quickly. -1

> Example:

Reviewer: ```suggestion
For example:
```

> - datasource metric queries

Reviewer 3: ```suggestion
- data source metric queries
```

> - data source metric queries

Reviewer 2: ```suggestion
- Data source metric queries
```

> - alert notifications

Reviewer 2: ```suggestion
- Alert notifications
```

> This can be used to limit access to internal systems that the server Grafana runs on can access but that users of Grafana should not be able to access. This feature does not affect traffic from the Grafana users browser.

Reviewer 3: ```suggestion
You can limit access to internal systems, such as you can deny _user_ access to the Grafana server. This feature does not affect traffic from the a Grafana user’s browser.
```

Reviewer 3: Vet this for technical accuracy after my changes.

Author: That's not quite correct. Another hard section to nail down.

Perhaps:
`You can limit access to internal systems, such that a user through (for example) data source metric queries is unable to reach systems that the Grafana server itself would be able to reach. This feature does not affect traffic from a Grafana user's browser.`

Reviewer: How about: 
You can limit access to internal systems so that a user through data source metric queries is unable to reach systems that the Grafana server itself would be able to reach. This feature does not affect traffic from a Grafana user's browser.

> Note: While request security works with backend plugins, it is possible to create a backend plugin that bypasses this security.

Reviewer 3: ```suggestion
Note: Although request security works with backend plugins, it is possible to create a backend plugin that bypasses this security.
```

Reviewer 3: "While" here means "Although", correct? Asked another way, nothing is simultaneously happening, or is it?

Author: Correct, thanks.

> Note: Although request security works with backend plugins, it is possible to create a backend plugin that bypasses this security.

Reviewer: ```suggestion
Note: Although request security works with backend plugins, you can create a backend plugin that bypasses this security.
```

Reviewer 3: +1

Reviewer 2: I'm seriously tempted to add "it is possible" and other common wordy expressions to the style guide.

## IP and hostname blocking

> It's possible to limit requests based on hostname and/or ip addresses in two ways.

Reviewer 3: ```suggestion
It is possible to limit requests based on a hostname, an IP address, or both.
```
I removed "two ways" here to make this more future proof wrt doc bugs, and when in doubt, readers can indeed count.

Reviewer 3: Notice how I worked around that "and/or" construction.

Author: Nice, thank you.

> It is possible to limit requests based on a hostname, an IP address, or both.

Reviewer: ```suggestion
You can limit requests based on a hostname, an IP address, or both.
```

Reviewer 3: +1

### Deny list

> Any request to a host or IP on the deny-list will be blocked.

Reviewer 3: ```suggestion
Any request to a hostname or IP address on the deny-list is denied.
```

> Any request to a hostname or IP address on the deny-list is denied.

Reviewer 2: ```suggestion
Grafana blocks any request to a hostname or IP address on the deny list.
```

Reviewer 3: I used "deny"  because it matches "deny / allow list verbiage. Agree with passive --> active voice. Also, block does read better imo.

Reviewer 2: I feel like "The deny list denies stuff" doesn't add much information, so I try to vary verbs a bit. "The deny list blocks stuff that is on it" actually explains something.

### Allow list

> If there are at least one host or IP on the allow-list any request to hosts or IP:s not on the allow-list will be blocked.

Reviewer 3: ```suggestion
If there is at least one hostname or IP address on the allow-list, any request to a hostname or IP address that is not on the allow-list is blocked.
```

This seems confusing to me. Can we write two sentences:

"Any request to a hostname or IP address that is only on the allow-list is allowed."

"Any request to a hostname or IP address on the allow-list that is also on the deny-list is denied."

Author: No, that's no quite what I'm trying to say. How about this?

```markdown
Any request to a hostname or IP not on the allow list is blocked. The allow list is only activated if there is at least one hostname or IP in it.
```

Reviewer 3: The allow list is only activated if there is at least one hostname or IP address on it. Furthermore, any request to a hostname or IP address that is not on the allow list is denied.

Reviewer 3: Updated my suggestion. How about now?

Author: I think that works.

> The allow list is only activated if there is at least one hostname or IP address on it. Furthermore, any request to a hostname or IP address that is not on the allow list is denied.

Reviewer: ```suggestion
The allow list is activated when there is at least one hostname or IP address on it. Furthermore, any request to a hostname or IP address that is not on the allow list is denied.
```

Reviewer 3: I reserve "when" to handle time-based contexts. For that reason, -1.

Reviewer: Per @oddlittlebird , if should always have a "then". So, that is what I have been following. Let's discuss these in detail. Otherwise, we are all following different review styles based on what we like?

Reviewer 3: Here, the construction is "only if" rather than "if then".

Reviewer 2: Correct. I'm also not a fan of "furthermore," but the if/then pairing is something I stick with.

Out of curiosity, how does one activate a list? Perhaps we are talking about the wrong thing here. Would it be more correct to say, "If there is at least one entry on the list, then any request to a hostname or IP address not on the list is denied"?

Reviewer 3: @xlson from a minimal viable improvement perspective, pick one and we can go over this offline.

Author: @oddlittlebird I like that version, reads a bit better to me.

> ### Example

Reviewer: ```suggestion
Example:
```

Reviewer 3: For example:

```toml
[security.egress]

> # a list of hostnames or IP addresses separated by spaces for which outgoing requests will be blocked

Reviewer 4: ```suggestion
# A list of hostnames or IP addresses separated by spaces for which requests are blocked.
```

host_deny_list = supersecret.internal 192.168.1.10

> # a list of hostnames or IP addresses separated by spaces for which requests will be allowed, all other requests will be blocked

Reviewer 4: ```suggestion
# A list of hostnames or IP addresses separated by spaces for which requests are allowed. All other requests are blocked.
```

host_allow_list = prometheus.internal

```

## Drop headers and cookies

> It's possible to set a list of cookies or headers that will be dropped from outgoing requests.

Reviewer 3: ```suggestion
It is possible to set a list of cookies or headers that are to be dropped from outgoing requests.
```

> It is possible to set a list of cookies or headers that are to be dropped from outgoing requests.

Reviewer 2: ```suggestion
You can set a list of cookies or headers that are to be dropped from outgoing requests.
```

Example:

```toml
[security.egress]

> # a list of headers that will be stripped from outgoing datasource and alerting requests

Reviewer 4: ```suggestion
# A list of headers that are stripped from the outgoing data source and any alerting requests.
```
